File: src/logging.rs
use std::{cell::RefCell, net::SocketAddr, time::Instant};

use axum::{
    extract::{ConnectInfo, Request},
    http::HeaderValue,
    response::Response,
};
use serde_json::json;
use tokio::{task, task_local};
use uuid::Uuid;

use crate::config::get_config;

task_local! {
    pub static LOG_MESSAGE: RefCell<String>;
}

pub struct RequestLogContext {
    pub start: Instant,
    pub request_id: String,
    pub method: String,
    pub url: String,
    pub source_host: String,
    pub status_code: Option<u16>,
    pub log_level: Option<u16>,
    pub log_level_name: Option<String>,
}

impl RequestLogContext {
    pub fn new(req: &Request) -> Self {
        let forwarded_for = req
            .headers()
            .get("X-Forwarded-For")
            .and_then(|hv: &HeaderValue| hv.to_str().ok());
        RequestLogContext {
            start: Instant::now(),
            request_id: req
                .headers()
                .get("X-Request-ID")
                .and_then(|header| header.to_str().ok())
                .map(String::from)
                .unwrap_or_else(|| format!("metadata-service-{}", Uuid::new_v4())),
            method: req.method().to_string(),
            url: req.uri().to_string(),
            source_host: match forwarded_for {
                Some(forwarded_for) => forwarded_for.to_string(),
                None => req
                    .extensions()
                    .get::<ConnectInfo<SocketAddr>>()
                    .map(|ConnectInfo(addr)| addr.ip().to_string())
                    .unwrap_or("Unknown".to_string()),
            }
            .to_string(),
            status_code: None,
            log_level: None,
            log_level_name: None,
        }
    }

    pub fn add_response_fields(self: &mut Self, response: &Response) {
        self.status_code = Some(response.status().as_u16());
        (self.log_level, self.log_level_name) = match response.status().as_u16() {
            200..=299 => (Some(10), Some("INFO".to_string())),
            400..=499 => (Some(20), Some("WARNING".to_string())),
            _ => (Some(30), Some("ERROR".to_string())),
        };
    }
}

pub async fn log_request(log_message: String, request_log_context: &RequestLogContext) {
    let log_entry = json!({
        "@timestamp": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        "command": get_config().command,
        "error.stack": "",
        "host": get_config().docker_host_name,
        "message": log_message,
        "method": request_log_context.method,
        "url": request_log_context.url,
        "statusCode": request_log_context.status_code,
        "responseTime": request_log_context.start.elapsed().as_millis(),
        "serviceName": "metadata-service",
        "serviceVersion": get_config().commit_id,
        "schemaVersion": "v3",
        "xRequestId": request_log_context.request_id,
        "source_host": request_log_context.source_host,
        "thread": task::id().to_string(),
        "level": request_log_context.log_level,
        "levelName": request_log_context.log_level_name,
        "loggerName": "request_logger",
    });
    println!("{log_entry}");
}

#[macro_export]
macro_rules! log {
    ($msg:expr) => {
        use crate::logging::LOG_MESSAGE;
        LOG_MESSAGE.with(|s| *s.borrow_mut() = $msg);
    };
}

#[macro_export]
macro_rules! run_handler_with_log_context {
    ($req:expr, $next:expr) => {{
        use crate::logging::LOG_MESSAGE;
        use std::cell::RefCell;
        LOG_MESSAGE
            .scope(RefCell::new("Responded".to_string()), async move {
                let response = $next.run($req).await;
                let log_message = LOG_MESSAGE.get().borrow().to_string();
                (response, log_message)
            })
            .await
    }};
}

File: metadata_service/config/logging.py
import re
import sys
import uuid
import json
import logging
import datetime
from time import perf_counter_ns
from typing import Callable

from fastapi import Request
from contextvars import ContextVar

from metadata_service.config import environment


request_start_time: ContextVar[int] = ContextVar("request_start_time")
correlation_id: ContextVar[str] = ContextVar("correlation_id")
method: ContextVar[str] = ContextVar("method")
url: ContextVar[str] = ContextVar("url")
remote_host: ContextVar[str] = ContextVar("remote_host")
response_status: ContextVar[int] = ContextVar("response_status")
response_time_ms: ContextVar[int] = ContextVar("response_time_ms")


class MicrodataJSONFormatter(logging.Formatter):
    def __init__(self):
        self.host = environment.get("DOCKER_HOST_NAME")
        self.command = json.dumps(sys.argv)
        self.commit_id = environment.get("COMMIT_ID")

    def format(self, record: logging.LogRecord) -> str:
        stack_trace = ""
        if record.exc_info is not None:
            stack_trace = self.formatException(record.exc_info)
        return json.dumps(
            {
                "@timestamp": datetime.datetime.fromtimestamp(
                    record.created,
                    tz=datetime.timezone.utc,
                ).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
                + "Z",
                "command": self.command,
                "error.stack": stack_trace,
                "host": self.host,
                "message": record.getMessage(),
                "level": record.levelno,
                "levelName": record.levelname,
                "loggerName": record.name,
                "method": method.get(""),
                "responseTime": response_time_ms.get(""),
                "schemaVersion": "v3",
                "serviceName": "metadata-service",
                "serviceVersion": self.commit_id,
                "source_host": remote_host.get(""),
                "statusCode": response_status.get(""),
                "thread": record.threadName,
                "url": url.get(""),
                "xRequestId": re.sub(r"[^\w\-]", "", correlation_id.get("")),
            }
        )


def setup_logging(app, log_level=logging.INFO):
    logger = logging.getLogger()
    logger.setLevel(log_level)

    formatter = MicrodataJSONFormatter()

    stream_handler = logging.StreamHandler()
    stream_handler.setFormatter(formatter)
    logger.addHandler(stream_handler)

    @app.middleware("http")
    async def add_process_time_header(request: Request, call_next: Callable):
        request_start_time.set(perf_counter_ns())
        corr_id = request.headers.get("X-Request-ID", None)
        if corr_id is None:
            correlation_id.set("metadata-service-" + str(uuid.uuid1()))
        else:
            correlation_id.set(corr_id)
        method.set(request.method)
        url.set(str(request.url))
        client = request.client
        host = ""
        if client is not None:
            host = client.host
        remote_host.set(host)

        response = await call_next(request)

        response_time = int(
            (perf_counter_ns() - request_start_time.get()) / 1_000_000
        )
        response_time_ms.set(response_time)
        response_status.set(response.status_code)
        response.headers["X-Request-ID"] = correlation_id.get()
        logger.info("responded")
        return response
